### ISSUE TITLE
Tautological compare warnings?

### ISSUE THREAD

**Issue by poizan42** [dotnet/coreclr#99]
These warnings looks like they could be bugs

	In file included from /home/poizan/coreclr/src/pal/src/file/filetime.cpp:77:
	/home/poizan/coreclr/src/pal/../inc/safemath.h:739:24: warning: comparison of
		  unsigned expression >= 0 is always true [-Wtautological-compare]
					if(lhs >= 0) //first case
					   ~~~ ^  ~
	/home/poizan/coreclr/src/pal/src/file/filetime.cpp:900:39: note: in
		  instantiation of member function 'ClrSafeInt<unsigned long>::subtraction'
		  requested here
		bool isSafe = ClrSafeInt<UINT64>::subtraction(
										  ^
	In file included from /home/poizan/coreclr/src/pal/src/file/filetime.cpp:77:
	/home/poizan/coreclr/src/pal/../inc/safemath.h:416:29: warning: comparison of
		  unsigned expression < 0 is always false [-Wtautological-compare]
			return ((lhs ^ rhs) < 0);
					~~~~~~~~~~~ ^ ~
	/home/poizan/coreclr/src/pal/../inc/safemath.h:733:16: note: in instantiation of
		  member function 'ClrSafeInt<unsigned long>::IsMixedSign' requested here
				if(IsMixedSign(lhs, rhs)) //test for +/- combo
				   ^
	/home/poizan/coreclr/src/pal/src/file/filetime.cpp:900:39: note: in
		  instantiation of member function 'ClrSafeInt<unsigned long>::subtraction'
		  requested here
		bool isSafe = ClrSafeInt<UINT64>::subtraction(
										  ^
	2 warnings generated.

	In file included from /home/poizan/coreclr/src/pal/src/memory/heap.cpp:37:
	/home/poizan/coreclr/src/pal/../inc/safemath.h:416:29: warning: comparison of
		  unsigned expression < 0 is always false [-Wtautological-compare]
			return ((lhs ^ rhs) < 0);
					~~~~~~~~~~~ ^ ~
	/home/poizan/coreclr/src/pal/../inc/safemath.h:682:17: note: in instantiation of
		  member function 'ClrSafeInt<unsigned long>::IsMixedSign' requested here
				if(!IsMixedSign(lhs, rhs))
					^
	/home/poizan/coreclr/src/pal/src/memory/heap.cpp:241:30: note: in instantiation
		  of member function 'ClrSafeInt<unsigned long>::addition' requested here
		if (!ClrSafeInt<size_t>::addition(dwBytes,nSize,fullsize))
								 ^
	1 warning generated.

	In file included from /home/poizan/coreclr/src/pal/src/misc/utils.cpp:40:
	/home/poizan/coreclr/src/pal/../inc/safemath.h:546:28: warning: comparison of
		  unsigned expression < 0 is always false [-Wtautological-compare]
						if(lhs < 0)
						   ~~~ ^ ~
	/home/poizan/coreclr/src/pal/src/misc/utils.cpp:241:30: note: in instantiation
		  of member function 'ClrSafeInt<unsigned long>::multiply' requested here
		if (!ClrSafeInt<size_t>::multiply(length,sizeof(WCHAR),fullsize))
								 ^
	In file included from /home/poizan/coreclr/src/pal/src/misc/utils.cpp:40:
	/home/poizan/coreclr/src/pal/../inc/safemath.h:416:29: warning: comparison of
		  unsigned expression < 0 is always false [-Wtautological-compare]
			return ((lhs ^ rhs) < 0);
					~~~~~~~~~~~ ^ ~
	/home/poizan/coreclr/src/pal/../inc/safemath.h:496:21: note: in instantiation of
		  member function 'ClrSafeInt<unsigned long>::IsMixedSign' requested here
					if(!IsMixedSign(lhs, rhs))
						^
	/home/poizan/coreclr/src/pal/src/misc/utils.cpp:241:30: note: in instantiation
		  of member function 'ClrSafeInt<unsigned long>::multiply' requested here
		if (!ClrSafeInt<size_t>::multiply(length,sizeof(WCHAR),fullsize))
								 ^
	2 warnings generated.

**Comment by poizan42** [comment]
Hmm it's just because ClrSafeInt is instantiated with an unsigned type. They should probably be silenced though - would it be a good idea to simply silence them right at the implementation in safemath.h? Maybe creating a macro for that purporse.

**Comment by janvorli** [comment]
@poizan42 I am curious, are you using a newer version of clang than 3.5? With 3.5, I am not getting these warnings.

**Comment by poizan42** [comment]
Interesting. No I'm using 3.2:

    clang --version
    Debian clang version 3.2-11 (tags/RELEASE_32/final) (based on LLVM 3.2)
    Target: x86_64-pc-linux-gnu
    Thread model: posix

**Comment by janvorli** [comment]
We only support clang 3.5 at the moment.

**Comment by poizan42** [comment]
Hmm seems the behaviour in clang was changed in http://llvm.org/bugs/show_bug.cgi?id=8682, it is not produced when I try with 3.5 myself either.

Is it documented anywhere that clang 3.5 is required? (There is also some issues with a few of the pal c_runtime tests because clang 3.2 does not support unicode string literals in C files but only C++)

**Comment by jkotas** [comment]
Should be addressed by #132
